packages/nextjs/src/server/authMiddleware.ts

```diff
@@ -0,0 +1,190 @@
+  req: NextRequest,
+  evt: NextFetchEvent,
+) => NextMiddlewareResult | Promise<NextMiddlewareResult>;
+
+type AuthMiddlewareParams = WithAuthOptions & {
+  /**
+   * A function that is called before the authentication middleware is executed.
+   * If a redirect response is returned, the middleware will respect it and redirect the user.
+   * If false is returned, the auth middleware will not execute and the request will be handled as if the auth middleware was not present.
+   */
+  beforeAuth?: BeforeAuthHandler;
+  /**
+   * A function that is called after the authentication middleware is executed.
+   * This function has access to the auth object and can be used to execute logic based on the auth state.
+   */
+  afterAuth?: AfterAuthHandler;
+  /**
+   * A list of routes that should be accessible without authentication.
+   * You can use glob patterns to match multiple routes or a function to match against the request object.
+   * Path patterns and regular expressions are supported, for example: `['/foo', '/bar(.*)'] or `[/^\/foo\/.*$/]`
+   * The sign in and sign up URLs are included by default, unless a function is provided.
+   * For more information, see: https://clerk.com/docs
+   */
+  publicRoutes?: RouteMatcherParam;
+  /**
+   * A list of routes that should be ignored by the middleware.
+   * This list typically includes routes for static files or Next.js internals.
+   * For improved performance, these routes should be skipped using the default config.matcher instead.
+   */
+  ignoredRoutes?: IgnoredRoutesParam;
+  /**
+   * Enables extra debug logging.
+   */
+  debug?: boolean;
+};
+
+export interface AuthMiddleware {
+  (): NextMiddleware;
+  (params: AuthMiddlewareParams): NextMiddleware;
+}
+
+const authMiddleware: AuthMiddleware = (...args: unknown[]) => {
+  const [params = {}] = args as [AuthMiddlewareParams?];
+  const { beforeAuth, afterAuth, publicRoutes, ignoredRoutes, ...options } = params;
+
+  const isIgnoredRoute = createRouteMatcher(ignoredRoutes || DEFAULT_IGNORED_ROUTES);
+  const isPublicRoute = createRouteMatcher(withDefaultPublicRoutes(publicRoutes));
+  const defaultAfterAuth = createDefaultAfterAuth(isPublicRoute);
+
+  return async (req: NextRequest, evt: NextFetchEvent) => {
+    if (isIgnoredRoute(req)) {
+      console.warn(receivedRequestForIgnoredRoute(req.nextUrl.href, JSON.stringify(DEFAULT_CONFIG_MATCHER)));
+      return setHeader(NextResponse.next(), constants.Headers.AuthReason, 'ignored-route');
+    }
+
+    const beforeAuthRes = await (beforeAuth && beforeAuth(req, evt));
+
+    if (beforeAuthRes === false) {
+      return setHeader(NextResponse.next(), constants.Headers.AuthReason, 'skip');
+    } else if (beforeAuthRes && isRedirect(beforeAuthRes)) {
+      return setHeader(beforeAuthRes, constants.Headers.AuthReason, 'redirect');
+    }
+
+    const requestState = await authenticateRequest(req, options);
+    if (requestState.isUnknown) {
```

Comment:
Hi team clerk.
I would really like someone to take a moment to answer me when does part 122->125 happen.